### PR TITLE
Add de.klayout.KLayout

### PR DESCRIPTION
**Pull request by willsALMANJ**
This PR adds packaging for KLayout (https://www.klayout.de/, https://github.com/KLayout/klayout). The maintainer has been asked a couple times about packaging as a Flatpak but seems uninterested (https://github.com/KLayout/klayout/issues/125).

### CODE REVIEW

File: de.klayout.KLayout.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  - install -Dm644 de.klayout.KLayout.appdata.xml ${FLATPAK_DEST}/share/appdata/${FLATPAK_ID}.appdata.xml
+  - install -Dm644 etc/klayout.desktop ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
+  - install -Dm644 etc/logo.png ${FLATPAK_DEST}/share/icons/hicolor/512x512/apps/${FLATPAK_ID}.png
+  - desktop-file-edit --set-icon=${FLATPAK_ID} /app/share/applications/${FLATPAK_ID}.desktop
+  sources:
+  - type: archive
+    url: https://www.klayout.org/downloads/source/klayout-0.26.5.tar.gz
+    sha256: 3f58be531b5570e9d01ac8077a84b90b91fe181e01dd4acc505c8a77dcdc8d88
+  - type: file
+    path: de.klayout.KLayout.appdata.xml
```
Comment (LeandroStanger) [inline de.klayout.KLayout.yaml:41]: New appdata files should use metainfo instead. https://www.freedesktop.org/software/appstream/docs/chap-Quickstart.html

**Comment by LeandroStanger** [comment]
bot, build de.klayout.KLayout

File: de.klayout.KLayout.yaml (hunk 1 above)
Comment (willsALMANJ) [reply de.klayout.KLayout.yaml:41]: Okay, I switched out the appdata file for a metainfo file.

**Comment by willsALMANJ** [comment]
Hmm, looking at the build pipeline -- I don't know that anyone has ever tried to build KLayout for ARM, but I guess it is okay if the build is successful.

**Comment by bilelmoussaoui** [comment]
How did you ended up choosing de.klayout.Klayout as an app-id?

File: de.klayout.KLayout.metainfo.xml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>GPL-3.0-or-later</project_license>
+  
+  <description>
+    <p>
+      Mask layout viewer and editor for the chip design engineer.
+    </p>
+  </description>
+  
+  <launchable type="desktop-id">org.zotero.KLayout.desktop</launchable>
```
Comment (bilelmoussaoui) [inline de.klayout.KLayout.metainfo.xml:17]: That doesn't look right?

Comment (willsALMANJ) [reply de.klayout.KLayout.metainfo.xml:17]: Oops, I copied that over from the appdata file which I modeled on the Zotero flatpak. I will fix it after we resolve the app ID question.

**Comment by willsALMANJ** [comment]
I went with de.klayout.KLayout because the KLayout project owns the klayout.de web domain, but I can change it to something else if that is better.

**Comment by bilelmoussaoui** [comment]
I just noticed that the mentioned domain name on the github repository https://github.com/KLayout/klayout is klayout.org, but it seems it just redirects to the klayout.de.

File: de.klayout.KLayout.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  sources:
+  - type: archive
+    url: https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.1.tar.gz
+    sha256: d418483bdd0000576c1370571121a6eb24582116db0b7bb2005e90e250eae418
+- name: klayout
+  buildsystem: simple
+  build-commands:
+  - bash ./build.sh -bin ${FLATPAK_DEST}/bin -j$FLATPAK_BUILDER_N_JOBS
+  post-install:
+  - install -Dm644 de.klayout.KLayout.metainfo.xml ${FLATPAK_DEST}/share/appdata/${FLATPAK_ID}.metainfo.xml
```
Comment (bilelmoussaoui) [inline de.klayout.KLayout.yaml:32]:
```suggestion
  - install -Dm644 de.klayout.KLayout.metainfo.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml
```
appdata is the old install directory

File: de.klayout.KLayout.metainfo.xml [hunk 4]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bilelmoussaoui) [inline de.klayout.KLayout.metainfo.xml:1]: The metainfo should be contributed upstream if possible

Comment (willsALMANJ) [reply de.klayout.KLayout.metainfo.xml:1]: Okay, I submitted it here: https://github.com/KLayout/klayout/pull/562

**Comment by willsALMANJ** [comment]
I think I addressed the feedback. I had a couple questions about the packaging:

* How does flatpak determine the version of the application? I didn't specify it any where. Is it pulling it from the tarball file name?
* KLayout can use Ruby for scripting, so I bundled in Ruby "by hand." Is that the best way to include Ruby? I didn't see a shared module or extension for Ruby support. Or should I have specified org.kde.Sdk as the runtime since that includes Ruby?

**Comment by nedrichards** [comment]
@willsALMANJ version as displayed to the user comes from the appdata version tags. Flatpak itself uses ostree to look at the difference between commits - you can have a read about that here: https://docs.flatpak.org/en/latest/under-the-hood.html

Best to just include Ruby rather than making the user get the SDK as well unless you generally need a bunch of developer tools.

File: de.klayout.KLayout.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+- --socket=wayland
+- --filesystem=host
+- --device=dri
+modules:
+- name: gdbm
+  config-opts:
+  - --enable-libgdbm-compat
+  sources:
+  - type: archive
+    url: ftp://ftp.gnu.org/gnu/gdbm/gdbm-1.18.1.tar.gz
```
Comment (barthalion) [inline de.klayout.KLayout.yaml:18]: Can you switch it to https?

File: de.klayout.KLayout.yaml [hunk 6]
```diff
@@ -0,0 +1,41 @@
+app-id: de.klayout.KLayout
```
Comment (barthalion) [inline de.klayout.KLayout.yaml:1]: Please reformat file as described in https://github.com/flathub/flathub/wiki/YAML-Style-Guide

Comment (willsALMANJ) [reply de.klayout.KLayout.yaml:1]: The file has been reformatted.

**Comment by willsALMANJ** [comment]
bot, build de.klayout.KLayout

**Comment by willsALMANJ** [comment]
bot, build de.klayout.KLayout

File: de.klayout.KLayout.yaml (hunk 5 above)
Comment (willsALMANJ) [reply de.klayout.KLayout.yaml]: I switched it. The gdbm site doesn't have an https link, but switching to https worked.

**Comment by willsALMANJ** [comment]
@nedrichards In the metainfo.xml file that I am including (and previously in the appdata file), there is a "versions" section but not a "this version" section, so is it defaulting to using the latest version? Or am I missing somewhere else where it looks for the version?

**Comment by barthalion** [comment]
Yes, it uses the latest version specified there.

**Comment by willsALMANJ** [comment]
What is the best way to resolve the failure? I switched everything from "appdata" to "metainfo" following [this comment](https://github.com/flathub/flathub/pull/1535#discussion_r428932095) above . Was I supposed to use the metainfo format rules but leave things named "appdata" or something else?

**Comment by barthalion** [comment]
The path is fine and the error is legit. It needs to provide `url` and `content_rating` tags.

**Comment by willsALMANJ** [comment]
bot, build de.klayout.KLayout

**Pull request closed without merging**